### PR TITLE
Add PMIX_TIMEOUT support for PMIx_Get and PMIx_Get_nb

### PR DESCRIPTION
**Pull request by igor-ivanov**
This change relates issue https://github.com/pmix/master/issues/27

@rhc54 could you look at this change and comment.
I think that it is a step to support PMIX_TIMEOUT option at least on pmix library layer.

### CODE REVIEW

**Comment by rhc54** [comment]
I think it's an okay first step - I think long-term we'll bring over the opal_hotel class to more generally support timeout operations. The only concern I have is that we now need to remove that pmix_timeout value from the array we pass up to the RM server, and also ensure that the pmix_server in the library doesn't use timeouts. Otherwise, we'll have multiple layers trying to "timeout" this operation, and a horrible race condition will be created.

In other words, we can only do the timeout at one level. I'm not sure if the server is better than the client for this purpose - we can look at that and discuss on the next call. Whichever level we select, we should _always_ do the timeout there to avoid mistakes.

**Comment by igor-ivanov** [comment]
@rhc54 completelly agree with your notices. I kept in mind the same thinks.
Just a note: I do not think we need remove pmix_info_t values passed by user as soon as we can ignore them on server side if it is required by a logic we would like to implement.

**Comment by igor-ivanov** [comment]
@rhc54 I am adding additional commit that relates wildcard logic. Could you think about it to discuss at meeting.

**Comment by rhc54** [comment]
@igor-ivanov I think not doing a dmodex for non-specific ranks is fine. Either we are looking for job-level info that should have been included in the initial handshake, or this request came thru a legacy interface and shouldn't have come until after a blocking fence with data exchange. Either way, it's an error.

Looking further at the question of where to do the timeout, I think the race conditions need to be thought through:

* if we put the timeout in the client, as you've proposed here, then the client must notify the pmix server (who will have to notify the RM) that the operation has been timed out so those levels can cancel whatever operation they have in-process. So we now have a multi-level race condition as the operation could have been completed at either the PMIx server or the RM level while the client hits the timeout, which means every level has to know what to do with a stale successful operation

* if we put the timeout in the pmix server, then we still have the race condition between the pmix server and the RM. We'll have to add RM functions that the pmix server can call to tell the RM that an operation has been timed out, and the RM and server will have to deal with stale successful operations.

* if we put the timeout in the RM, then only the RM has to deal with stale successful operations, and the rest of the race conditions disappear. The negative here is that we are dependent upon the RM to correctly support timeout directives.

We can discuss the options tomorrow, though I confess I'm leaning towards the latter as it is the less complex logic. RMs are _always_ free to ignore a non-required info key, and to return an error if a required info key isn't supported. So long as it is easy for a user to make the required/optional choice clear, it seems to me that this might be our best option.

Either way, we need to clearly explain how to declare an info key as "required".

**Comment by igor-ivanov** [comment]
@rhc54 thank you for sharing your opinion.
a couple of words here to save time on the meeting.
* I think not doing a modex for non-specific ranks is a way to go but there is a bad side of this.
Example:
 - rank Y puts KEY_A to PMIX_GLOBAL
 - rank X know that someone put KEY_A to PMIX_GLOBAL
 - PMIx_Fence is called with data collection
 - rank X calls PMIx_Get(KEY_A, PMIX_UNDEF) but __does not get KEY_A__ (because of proposed logic)
Do you think the same or I missed something?

* I selected timeout on client side as simplest way and a way when server must not know about any timeout logic.
 - if PMIx_Get exits by TIMEOUT and client knows about it checking status
 - server continues processing a GET request w/o any information about TIMEOUT and does all work as usual
 - _getnb_cbfunc() does all it should do excluding callback() call
- What we have as a result: 
  - a client gets TIMEOUT status and is able to change it`s logic taking into account this information;
  - a server processed request as usual but it is not a crime;

**Comment by rhc54** [comment]
I don't think your example for the modex issue is correct. If you do a
Fence with data collection, then all the data that has been published is
available to you. Thus, if you call Get and the data isn't found, then
nobody has published it and even a direct modex will fail. This is why your
proposed logic is okay - in the use-case where PMIX_UNDEF is used, we are
talking about legacy applications and that means data will always have been
collected.

I actually believe that client-level timeout is the most complex, not the
simplest, to implement. You have only dealt with the client-level code.
There is a substantial amount of code required in the server and RM levels
as well.

Remember, we are defining a policy for _all_ operations that can include a
timeout, not just Get. Spawn, lookup, etc. all involve RM operations that
must be cancelled if they timeout. You can't just let those hang out there
in limbo if the client library times out as the client may choose to
request an alternative action that could conflict - e.g., request spawn of
a smaller version of the application that would use some of the same
resources.

So the request to "cancel" the operation has to be communicated upward,
which is why there is a race condition.

**Comment by igor-ivanov** [comment]
@rhc54 sorry for changing your vision about current fence logic but in reality my example is correct.
Server side does not share modex info with clients on fence now. See https://github.com/pmix/master/commit/f21d47962f126709d756feab4f2203b1c612ed60
After talking with @artpol84 I am looking at a way to control this logic using pmix_info_t value.

**Comment by rhc54** [comment]
I realize that the data stops at the daemon. My point is that a "direct
modex" call to retrieve the data from another remote daemon won't happen.
You will execute the local fetch from the local daemon, just like we always
do. If the server finds that no data has arrived for that proc, then it
needs to return an error in that scenario (i.e., when a fence with data
collection has occurred on that job).

HTH

**Comment by igor-ivanov** [comment]
there is a decision to do this using another way.
